osu.Game.Modes.Taiko/Scoring/TaikoScoreProcessor.cs

```diff
@@ -0,0 +1,270 @@
+        {
+            double hpMultiplierNormal = 1 / (hp_hit_great * beatmap.HitObjects.FindAll(o => o is Hit).Count * BeatmapDifficulty.DifficultyRange(beatmap.BeatmapInfo.Difficulty.DrainRate, 0.5, 0.75, 0.98));
+
+            hpIncreaseTick = hp_hit_tick;
+            hpIncreaseGreat = hpMultiplierNormal * hp_hit_great;
+            hpIncreaseGood = hpMultiplierNormal * BeatmapDifficulty.DifficultyRange(beatmap.BeatmapInfo.Difficulty.DrainRate, hp_hit_good_max, hp_hit_good, hp_hit_good);
+            hpIncreaseMiss = BeatmapDifficulty.DifficultyRange(beatmap.BeatmapInfo.Difficulty.DrainRate, hp_miss_min, hp_miss_mid, hp_miss_max);
+
+            var finishers = beatmap.HitObjects.FindAll(o => o is Hit && o.IsFinisher);
+
+            // This is a linear function that awards:
+            // 10 times bonus points for hitting a finisher with both keys with 30 finishers in the map
+            // 3 times bonus points for hitting a finisher with both keys with 120 finishers in the map
+            finisherScoreScale = -7d / 90d * MathHelper.Clamp(finishers.Count, 30, 120) + 111d / 9d;
+
+            foreach (var obj in beatmap.HitObjects)
+            {
+                if (obj is Hit)
+                {
+                    AddJudgement(new TaikoJudgement
+                    {
+                        Result = HitResult.Hit,
+                        TaikoResult = TaikoHitResult.Great,
+                        SecondHit = obj.IsFinisher
+                    });
+                }
+                else if (obj is DrumRoll)
+                {
+                    for (int i = 0; i < ((DrumRoll)obj).TotalTicks; i++)
+                    {
+                        AddJudgement(new TaikoDrumRollTickJudgement
+                        {
+                            Result = HitResult.Hit,
+                            TaikoResult = TaikoHitResult.Great,
+                            SecondHit = obj.IsFinisher
+                        });
+                    }
+
+                    AddJudgement(new TaikoJudgement
+                    {
+                        Result = HitResult.Hit,
+                        TaikoResult = TaikoHitResult.Great,
+                        SecondHit = obj.IsFinisher
+                    });
+                }
+                else if (obj is Bash)
+                {
+                    AddJudgement(new TaikoJudgement
+                    {
+                        Result = HitResult.Hit,
+                        TaikoResult = TaikoHitResult.Great
+                    });
+                }
+            }
+
+            maxTotalHits = totalHits;
+            maxComboPortion = comboPortion;
+        }
+
+        protected override void UpdateCalculations(TaikoJudgement newJudgement)
+        {
+            var tickJudgement = newJudgement as TaikoDrumRollTickJudgement;
+
+            // Don't consider ticks as a type of hit that counts towards map completion
+            if (tickJudgement == null)

```

Comment:
i'd also restructure all the code below to be

```
if (isTick)
{
    //do everything tick related
}
else
{
    //everything else
}
```

i think it will simplify things. you may even be able to do an early return for ticks, remove the else{} and reduce nesting